library/alloc/src/vec/is_zero.rs

```diff
@@ -147,6 +147,24 @@ impl_is_zero_option_of_nonzero!(
     NonZeroIsize,
 );
 
+macro_rules! impl_is_zero_option_of_num {
+    ($($t:ty,)+) => {$(
+        unsafe impl IsZero for Option<$t> {
+            #[inline]
+            fn is_zero(&self) -> bool {
+                core::intrinsics::assert_zero_valid::<Option<$t>>();
+                if let Some(n) = self {
+                    *n == 0
```

Comment:
This is not correct unless the `Some` tag is also 0 -- and it probably isn't!

```rust
fn main() {
    let opt = Some(0u8);
    let bytes: [u8; 2] = unsafe { core::mem::transmute(opt) };
    dbg!(bytes);
}
```

```
[src/main.rs:4] bytes = [
    1,
    0,
]
```

I think it was ok when you had this just for `None` -- then the value memory is indeterminate, but fine to treat as zero for the purpose of the new allocation.

Reply:
Agreed; I think this should just be `self.is_none()`.

Reply:
Ah, I wasn't quite sure that was right. Thanks for spotting!